### ISSUE TITLE
Integration with autorest?

### ISSUE THREAD

**Issue by SbiCA** [aspnet/HttpClientFactory#54]
Hi guys,

@rynowak cool demo at the community standup.

I was wondering if there is chance to see a an example for [Autorest](https://github.com/Azure/autorest)?

Because I often end up in introducing a clientfactory for my autorest clients especially with OAuth.

```csharp
   public class ApiClientFactory : IApiClientFactory
    {
        private readonly string _authenticationServiceBaseUrl;
        private readonly string _apiBaseUrl;

        public ApiClientFactory (string authenticationServiceBaseUrl, string apiBaseUrl)
        {
            _authenticationServiceBaseUrl = authenticationServiceBaseUrl;
            _apiBaseUrl = apiBaseUrl;
        }

        public IPortalApiClient GetApiClient(AuthenticationCredentials credentials)
        {
            // credentials contain client id, secret, scopes requested ...
            var accesstokenProvider = new AccessTokenProvider(_authenticationServiceBaseUrl,credentials);
            return new ApiClient(new Uri(_apiBaseUrl),new TokenCredentials(accesstokenProvider));
        }
    }
```

**Comment by glennc** [comment]
I haven't used AutoRest much yet, but I am good with us having a sample. I'll have a look at what they are doing and how we can plug in appropriately.

**Comment by rynowak** [comment]
@glennc - not urgent but on your plate

**Comment by robertmclaws** [comment]
Hello, I'm the developer of [PortableRest](https://github.com/CloudNimble/PortableRest), another library to help REST client authors build simple APIs for their developers. I'm really interested in implementing this in PortableRest so that people don't run into these lifetime issues with APIs built on my helper.

Most of the samples I've seen for using this library assume it will be used directly in a web app, and require the use Microsoft's DI implementation. Would be great if you could post a sample of how to use the `HttpClientFactory` WITHOUT registering it in an existing DI container, so it can be used in libraries like mine.

Would also be great if the samples could show options like setting allowed compression schemes, using subclassed `HttpMessageHandlers`, passing in `CookieContainers`, etc.

Thanks!

**Comment by MichaCo** [comment]
Hi @glennc, @rynowak,

I'm heavily using autorest in our projects to access our many ASP.NET Core services.

To work around DNS and local port issues and to handle `HttpClientHandler` lifetime properly, I'm actually currently using parts of the code of this library (before it got released) and added things to expose the factory managed `HttpClientHandler` to make it work with autorest generated clients.

It works well, though, I'd like to replace my implementation with the "real" package one day...
I'd also love to use the Polly extensions and named client (which I haven't added to my current implementation to keep things simple..)

It is actually not very easy/friendly, to integrate autorest with the HttpClientFactory right now, mainly because autorest doesn't generate constructors taking `HttpClient` (otherwise I could just inject a named HttpClient).

I've created a small GitHub repository to play with the integration; maybe you could take a look and give me some hints of how to improve it? 

https://github.com/MichaCo/AutorestWithHttpClientFactory

Thanks!

**Comment by glennc** [comment]
@robertmclaws Having you instantiate and use the factory yourself without the DI system isn't something we designed for. You shouldn't need the factory just for lifetime though. The sockets based HttpClientHandler added in 2.1 has a PooledConnectionLifetime property that is supposed to take the place of the service point manager timeouts in .NET framework and allow connections to be re-created and get DNS updates. Does that make sense and resonate or do you still feel that you would want to use the factory in some way? If so what other benefits are you looking for?

**Comment by robertmclaws** [comment]
@glennc Thanks for the insight :). I think the team might be missing the value for the Factory pattern here in non-ASPNET situations. Let's say I have a long-running server-side batch process, and I'm using a 3rd-party API powered by AutoRest or PortableRest. There is _value_ in just being able to call `ApiClientFactory.GetApiClient()` (which under the covers calls `HttpClientFactory.GetHttpClient()` and have it give my customer a preconfigured HttpClient to use. That instance could be disposed when I'm done with the current batch request, and it does the right thing... the way HttpClient was SUPPOSED to be designed, and the way all of the current documentation tells me to use it.

Putting it into a DI container is fine. Making the system dependent on other ASP.NET stuff is not. This code is a clear candidate for .NET Standard, but if you've designed it in such a way that it can't be used outside the context of ASP.NET... that's going to be a problem. And making MY OWN HttpClientFactory that implements PooledConnectionLifetime doesn't seem like the best idea either.

I should also point out that this is why it would have been great to come up with what the HttpClientFactory should do as a part of an open process, and not just build it internally without soliciting requirements from end users. I think that's part of the reason HttpClient went from "a great idea" to "a freakin mess" in the first place.

Happy to help out any way I can...

**Comment by davidfowl** [comment]
> Making the system dependent on other ASP.NET stuff is not.

I do not see a world where components we build don't depend on other components we build. We built them so that we don't have to keep rebuilding other abstractions (logging, configuration, DI etc).

If you don't care for the DI container treat it like an implementation detail. If we were to make a wrapper API it would *still* use the DI container internally for itself (and the logging interfaces). Making a second configuration API is a non starter IMO as we have no plans to design an API that needs to be used.

Most people seem to *just* want to pooling anyways and if that's really the overwhelming majority wants then we need to fix HttpClient itself.

**Comment by robertmclaws** [comment]
@davidfowl That's not what I'm talking about. You missed the previous statement where I said the DI would be fine. DI is not an ASP.NET host-specific thing. But if you look at the end of @MichaCo's example, you can't currently use HttpClientFactory outside of the ASP.NET host. And that's what I'm talking about.

**Comment by martincostello** [comment]
That's definitely not true, because I've written unit tests for extension code for `IHttpClientFactory`, and they're not in an ASP.NET Core host, they're just xunit unit tests.

**Comment by robertmclaws** [comment]
Check out https://github.com/MichaCo/AutorestWithHttpClientFactory/blob/master/src/Runner/Program.cs#L70

**Comment by martincostello** [comment]
That's interesting, considering there's [no obvious usages of it in the repo](https://github.com/aspnet/HttpClientFactory/search?q=IHttpContextAccessor&type=Code), _but_ hosting vs. the presence of a dependency are two different things. If it does need that to be there that's less optimal, but then that's just a slightly unfortunate NuGet reference lying around if it was a console app, for example.

**Comment by davidfowl** [comment]
Nothing is http client factory depends on ASP.NET Core. Looks like you’re having an unrelated problem.

**Comment by davidfowl** [comment]
The problem is here https://github.com/MichaCo/AutorestWithHttpClientFactory/blob/21462ef8d7edef3a65cd795ebbf850e0d404657e/src/Runner/Program.cs#L60 and https://github.com/MichaCo/AutorestWithHttpClientFactory/blob/21462ef8d7edef3a65cd795ebbf850e0d404657e/src/Runner/Program.cs#L79.

You expecting a dependency that hasn't been registered to be in the container

**Comment by martincostello** [comment]
I don't know about docs, but off the top of my head you can do something like this to get yourself a factory:

```csharp
var services = new ServiceCollection();
services.AddHttpClient();

var provider = services.BuildServiceProvider();
var factory = provider.GetRequiredService<IHttpClientFactory>();
var client = factory.CreateHttpClient();
```

**Comment by MichaCo** [comment]
@davidfowl The examples I wrote are basically mocking some things a normal (mico) service would do, think about ConfigureServices in a console app.

Sure, you can use HttpClientFactory without any ASP.NET Hosting, but you still need DI/ServiceCollection to get a factory. Meaning, there is at least some dependency on DI.
I don't mind though, just saying ^^

**Comment by davidfowl** [comment]
Yep thats by design.

**Comment by abatishchev** [comment]
What would help me to use it outside an ASP.NET Core app (actually calling an ASP.NET Core app) is decreased amount of a boiler plate code, like what @martincostello posted. Less is better. Amd that's exactly what we have withing an app but not outside, yet.

**Comment by davidfowl** [comment]
So these 5 lines are too much boilerplate:

```C#
var services = new ServiceCollection();
services.AddHttpClient();
var provider = services.BuildServiceProvider();
var factory = provider.GetRequiredService<IHttpClientFactory>();
var client = factory.CreateHttpClient();
```

**Comment by abatishchev** [comment]
2 is already too much. Because it's boilerplate. Would be cool if I there would be an extension method, etc. which would hide it from me.

**Comment by stevejgordon** [comment]
@abatishchev It might not be quite what you're looking for but I'm currently using IHttpClientFactory in a .netstandard library intended to be used in a .NET Core microservice to call an ASP.NET Core API. The .NET Coremicroservice is a .NET Core console app and I'm using the Generic Host and IHostedService which means registering using the existing DI extensions all works fine (so far)! I plan to blog about the approach in the future as I tackle using this in a real system.

**Comment by glennc** [comment]
@MichaCo this thing would solve a chunk of your autorest problem right? https://github.com/aspnet/HttpClientFactory/issues/110

**Comment by davidfowl** [comment]
> 2 is already too much. Because it's boilerplate. Would be cool if I there would be an extension method, etc. which would hide it from me.

So you just want to use the factory for lifetime management, but not any of the other features.

**Comment by MichaCo** [comment]
@glennc yes, waiting what'll happen with that ;)

**Comment by robertmclaws** [comment]
Well, just to be clear, without documentation on how to do it, yes 5 lines is too much. Most people are just going to want to call `HttpClientFactory.GetHttpClient()` and expect it to just work. 

TBH, there should be a static overload that abstracts the DI container stuff completely away from the end user. Because those first 4 lines could easily be called internally if the user has not registered the factory with a DI container by the time an HttpClient is requested.

**Comment by abatishchev** [comment]
>Well, just to be clear, without documentation on how to do it, yes 5 lines is too much. Most people are just going to want to call HttpClientFactory.GetHttpClient() and expect it to just work.

Exactly my point. First when just I read about this new feature I tried to use it. And gave up once realized it comes form services what means ASP.NET Core app context. And only today, during this conversation, it turned out to be constructable/usable manually. Because the documentation say nothing about this possibility.

>TBH, there should be a static overload that abstracts the DI container stuff completely away from the end user. Because those first 4 lines could easily be called internally if the user has not registered the factory with a DI container by the time an HttpClient is requested.

I'm registering the resulting HttpClient withing my own Simple Injector container. Am I doing this wrong?

**Comment by davidfowl** [comment]
> Most people are just going to want to call HttpClientFactory.GetHttpClient() and expect it to just work.

Sorry, we just don't design APIs that way. I'm extremely hesitant to have more than one configuration API. This isn't a BCL API, unfortunately the name makes it seem that way so we set expectations incorrectly here. Our entire design (especially typed clients) revolve around having something that can activate your typed client and hand you dependencies on demand, and the DI container we have is *great* for that so it's not going anywhere. Building on these primitives that we've been working on for several years now makes it trivially possibly to do things like wire up logging to *any* provider that supports Microsoft.Extensions.Logging without any extra glue or extra packages. That's the value in building on these primitives (those benefits multiply when you write components or message handlers that require dependencies as well).

> I'm registering the resulting HttpClient withing my own Simple Injector container. Am I doing this wrong?

You want to put the factory in there, not the resulting HttpClient. 

> Well, just to be clear, without documentation on how to do it, yes 5 lines is too much. 

> Exactly my point. First when just I read about this new feature I tried to use it. And gave up once realized it comes form services what means ASP.NET Core app context. And only today, during this conversation, it turned out to be constructable/usable manually. Because the documentation say nothing about this possibility.

@glennc @rynowak and I discussed that yesterday and will write docs on how to use it outside of an ASP.NET Core application.
